tsdb/engine/tsm1/reader_test.go

```diff
@@ -977,6 +977,95 @@ func TestTSMReader_File_Read(t *testing.T) {
 		t.Fatalf("read values count mismatch: exp %v, got %v", exp, got)
 	}
 }
+
+func TestValidate(t *testing.T) {
+	dir := MustTempDir()
+	defer os.RemoveAll(dir)
+	f := MustTempFile(dir)
+	defer f.Close()
+
+	w, err := tsm1.NewTSMWriter(f)
+	if err != nil {
+		t.Fatalf("unexpected error creating writer: %v", err)
+	}
+
+	var data = []struct {
+		key    string
+		values []tsm1.Value
+	}{
+		{"float", []tsm1.Value{
+			tsm1.NewValue(time.Unix(1, 0), 1.0)},
+		},
+		{"int", []tsm1.Value{
+			tsm1.NewValue(time.Unix(1, 0), int64(1))},
+		},
+		{"bool", []tsm1.Value{
+			tsm1.NewValue(time.Unix(1, 0), true)},
+		},
+		{"string", []tsm1.Value{
+			tsm1.NewValue(time.Unix(1, 0), "foo")},
+		},
+	}
+
+	for _, d := range data {
+		if err := w.Write(d.key, d.values); err != nil {
+			t.Fatalf("unexpected error writing: %v", err)
+		}
+	}
+
+	if err := w.WriteIndex(); err != nil {
+		t.Fatalf("unexpected error writing index: %v", err)
+	}
+
+	if err := w.Close(); err != nil {
+		t.Fatalf("unexpected error closing: %v", err)
+	}
+
+	f, err = os.Open(f.Name())
+	if err != nil {
+		t.Fatalf("unexpected error open file: %v", err)
+	}
+
+	r, err := tsm1.NewTSMReaderWithOptions(
+		tsm1.TSMReaderOptions{
+			MMAPFile: f,
+		})
+	if err != nil {
+		t.Fatalf("unexpected error created reader: %v", err)
+	}
+	mmapCheckSum := r.ValidateChecksum()
+	r.Close()
+
+	for key, checksums := range mmapCheckSum {
+		for checksum, valid := range checksums {
+			if !valid {
+				t.Fatalf("Expected for key %s with a valid checksum %s, got: %t", key, checksum, valid)
```

Comment:
It's preferable to write test failure messages showing what you got and then what you expected. One way might be:

 ```go
if got, exp := valid, true; got != exp {
    t.Fatalf("got %v but expected %v for key %s with checksum %s", got, exp, key, checksum)
}
```